File: apps/client-macos/Sources/TimeTrack/Capture/ScreenshotUploader.swift
import Foundation

protocol ScreenshotUploading {
    func upload(id: String, capturedAt: Date, group: CaptureGroup?, jpeg: Data) async -> UploadResult
}

/// PRD §7.4 — uploads one captured screenshot to `<baseURL>/screenshots` as multipart/form-data.
/// The `id` + `timestamp` TEXT fields MUST precede the file part (2.2a §5): `@fastify/multipart`'s
/// `req.file()` only exposes fields parsed before the file, so a file-first body yields undefined
/// metadata → 422 on every upload. `userId` is the session (server reads the token, never a field).
/// Idempotent upsert on the client-minted `id`, so a lost-201 retry is a no-op. Reuses `UploadResult`
/// + `TimeEntryUploader.classify`. Not a capture path — no AckGate.
final class ScreenshotUploader: ScreenshotUploading {
    private let baseURL: URL
    private let session: AuthSession

    init(baseURL: URL, session: AuthSession) {
        self.baseURL = baseURL
        self.session = session
    }

    func upload(id: String, capturedAt: Date, group: CaptureGroup?, jpeg: Data) async -> UploadResult {
        let timestampISO = Self.iso.string(from: capturedAt)
        do {
            let token = try await session.accessToken()
            let status = try await post(id: id, timestampISO: timestampISO, group: group, jpeg: jpeg, token: token)
            if status == 401 {
                let refreshed = try await session.forceRefresh()
                let retried = try await post(id: id, timestampISO: timestampISO, group: group, jpeg: jpeg, token: refreshed)
                return TimeEntryUploader.classify(status: retried)
            }
            return TimeEntryUploader.classify(status: status)
        } catch {
            return .transient
        }
    }

    /// Pure multipart builder — field order (all TEXT fields, then the file) is the 2.2a
    /// invariant: `req.file()` on the server only exposes fields parsed BEFORE the file part, so
    /// a file-first body yields undefined metadata and 422s every upload.
    static func multipartBody(boundary: String, id: String, timestampISO: String,
                              group: CaptureGroup?, jpeg: Data) -> Data {
        var body = Data()
        func boundaryLine() { body.append(Data("--\(boundary)\r\n".utf8)) }
        func field(_ name: String, _ value: String) {
            boundaryLine()
            body.append(Data("Content-Disposition: form-data; name=\"\(name)\"\r\n\r\n".utf8))
            body.append(Data("\(value)\r\n".utf8))
        }
        field("id", id)               // MUST come first
        field("timestamp", timestampISO)  // MUST precede the file
        if let group {
            // Optional on the server, so an API that predates multi-display capture ignores
            // these rather than rejecting the upload.
            field("captureGroupId", group.id)
            field("displayIndex", String(group.displayIndex))
            field("displayCount", String(group.displayCount))
        }
        boundaryLine()
        body.append(Data("Content-Disposition: form-data; name=\"file\"; filename=\"\(id).jpg\"\r\n".utf8))
        body.append(Data("Content-Type: image/jpeg\r\n\r\n".utf8))
        body.append(jpeg)
        body.append(Data("\r\n--\(boundary)--\r\n".utf8))
        return body
    }

    private func post(id: String, timestampISO: String, group: CaptureGroup?, jpeg: Data,
                      token: String) async throws -> Int {
        let boundary = "TimeTrack-\(UUID().uuidString)"
        var request = URLRequest(url: baseURL.appendingPathComponent("screenshots"))
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        request.httpBody = Self.multipartBody(boundary: boundary, id: id, timestampISO: timestampISO,
                                              group: group, jpeg: jpeg)
        let (_, response) = try await URLSession.shared.data(for: request)
        return (response as? HTTPURLResponse)?.statusCode ?? 0
    }

    private static let iso: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()
}

File: apps/client-windows/src/NiftyTimer/Sync/ScreenshotUploader.cs
using System.Globalization;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using NiftyTimer.Auth;
using NiftyTimer.Storage;
using NiftyTimer.Tracking;

namespace NiftyTimer.Sync;

/// <summary>The screenshot-upload seam. Tests substitute a fake.</summary>
public interface IScreenshotUploading
{
    Task<UploadResult> UploadAsync(
        string id,
        DateTimeOffset capturedAt,
        CaptureGroup group,
        byte[] jpeg,
        CancellationToken cancellationToken = default);
}

/// <summary>
/// PRD §7.4 — uploads one captured screenshot to <c>&lt;baseUrl&gt;/screenshots</c> as
/// <c>multipart/form-data</c>.
///
/// **Every TEXT field must precede the file part.** <c>@fastify/multipart</c>'s <c>req.file()</c>
/// only exposes fields that were parsed BEFORE the file, so a file-first body arrives with
/// undefined metadata and 422s every single upload — which then classifies as permanent and
/// silently discards the image. The body is therefore built by hand rather than with
/// <c>MultipartFormDataContent</c>: a pure builder can have its field order asserted by a test,
/// where an ordering that lives in a series of <c>Add</c> calls cannot.
///
/// <c>userId</c> is never sent. The server attributes the upload to the bearer token, and an
/// extra field would be ignored at best.
///
/// The upload is idempotent on the client-minted <c>id</c>, so a retry after a lost 201 is a
/// no-op. Not a capture path — this sends what capture already produced — so no
/// <see cref="Policy.AckGate"/>.
/// </summary>
public sealed class ScreenshotUploader : IScreenshotUploading
{
    private readonly HttpClient _http;
    private readonly Uri _baseUrl;
    private readonly AuthSession _session;

    public ScreenshotUploader(HttpClient http, Uri baseUrl, AuthSession session)
    {
        _http = http;
        _baseUrl = baseUrl;
        _session = session;
    }

    public async Task<UploadResult> UploadAsync(
        string id,
        DateTimeOffset capturedAt,
        CaptureGroup group,
        byte[] jpeg,
        CancellationToken cancellationToken = default)
    {
        var timestamp = UuidV7.Iso(capturedAt);
        try
        {
            var token = await _session.AccessTokenAsync(cancellationToken).ConfigureAwait(false);
            var status = await PostAsync(id, timestamp, group, jpeg, token, cancellationToken).ConfigureAwait(false);
            if (status == (int)HttpStatusCode.Unauthorized)
            {
                var refreshed = await _session.ForceRefreshAsync(cancellationToken).ConfigureAwait(false);
                var retried = await PostAsync(id, timestamp, group, jpeg, refreshed, cancellationToken)
                    .ConfigureAwait(false);
                return TimeEntryUploader.Classify(retried); // a second 401 → AuthFailed
            }

            return TimeEntryUploader.Classify(status);
        }
        catch (Exception e) when (e is AuthException or NotAuthenticatedException or HttpRequestException
                                      or TaskCanceledException)
        {
            return new UploadResult.Transient();
        }
    }

    /// <summary>
    /// The pure body builder. Field order is the invariant, and it is what
    /// <c>ScreenshotUploaderTests</c> asserts: <c>id</c>, <c>timestamp</c>, then the grouping
    /// fields, then the file — never the other way round.
    ///
    /// The grouping fields are optional on the server so that a client predating multi-display
    /// capture is still accepted; this client always sends them.
    /// </summary>
    internal static byte[] MultipartBody(string boundary, string id, string timestampIso, CaptureGroup group, byte[] jpeg)
    {
        using var body = new MemoryStream();

        void Ascii(string text)
        {
            var bytes = Encoding.UTF8.GetBytes(text);
            body.Write(bytes, 0, bytes.Length);
        }

        void Field(string name, string value)
        {
            Ascii($"--{boundary}\r\n");
            Ascii($"Content-Disposition: form-data; name=\"{name}\"\r\n\r\n");
            Ascii($"{value}\r\n");
        }

        Field("id", id);               // MUST be first
        Field("timestamp", timestampIso); // MUST precede the file
        Field("captureGroupId", group.Id);
        Field("displayIndex", group.DisplayIndex.ToString(CultureInfo.InvariantCulture));
        Field("displayCount", group.DisplayCount.ToString(CultureInfo.InvariantCulture));

        Ascii($"--{boundary}\r\n");
        Ascii($"Content-Disposition: form-data; name=\"file\"; filename=\"{id}.jpg\"\r\n");
        Ascii("Content-Type: image/jpeg\r\n\r\n");
        body.Write(jpeg, 0, jpeg.Length);
        Ascii($"\r\n--{boundary}--\r\n");

        return body.ToArray();
    }

    private async Task<int> PostAsync(
        string id,
        string timestampIso,
        CaptureGroup group,
        byte[] jpeg,
        string token,
        CancellationToken cancellationToken)
    {
        var boundary = $"NiftyTimer-{Guid.NewGuid():N}";

        using var content = new ByteArrayContent(MultipartBody(boundary, id, timestampIso, group, jpeg));
        var contentType = new MediaTypeHeaderValue("multipart/form-data");
        contentType.Parameters.Add(new NameValueHeaderValue("boundary", boundary));
        content.Headers.ContentType = contentType;

        using var request = new HttpRequestMessage(HttpMethod.Post, new Uri(_baseUrl, "screenshots"))
        {
            Content = content,
        };
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

        using var response = await _http.SendAsync(request, cancellationToken).ConfigureAwait(false);
        return (int)response.StatusCode;
    }
}
